File: vkatz-lib/src/main/java/by/vkatz/utils/ActivityNavigator.java
package by.vkatz.utils;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;

import java.io.Serializable;

/**
 * Created by Katz on 17.06.2016.
 */

@SuppressWarnings({"unused", "WeakerAccess"})
public class ActivityNavigator<A extends Activity> {
    private A activity;
    private int flags = 0;
    private int animationIn = -1, animationOut = -1;
    private Bundle bundle;
    private Functions.Func1<Void, Intent> intentConfigurator = null;

    private ActivityNavigator(A activity) {
        this.activity = activity;
        bundle = new Bundle();
    }

    public static <A extends Activity> ActivityNavigator forActivity(A activity) {
        return new ActivityNavigator<>(activity);
    }

    public static Bundle getData(Activity activity) {
        try {
            return activity.getIntent().getExtras();
        } catch (Exception e) {
            return null;
        }
    }

    public ActivityNavigator noHistory() {
        flags |= Intent.FLAG_ACTIVITY_NO_HISTORY;
        return this;
    }

    public ActivityNavigator clearTop() {
        flags |= Intent.FLAG_ACTIVITY_CLEAR_TOP;
        return this;
    }

    public ActivityNavigator reorderToFront() {
        flags |= Intent.FLAG_ACTIVITY_REORDER_TO_FRONT;
        return this;
    }

    public ActivityNavigator animation(int animationIn, int animationOut) {
        this.animationIn = animationIn;
        this.animationOut = animationOut;
        return this;
    }

    public ActivityNavigator withData(Bundle data) {
        bundle.putAll(data);
        return this;
    }

    public ActivityNavigator withData(String key, String value) {
        bundle.putString(key, value);
        return this;
    }

    public ActivityNavigator withData(String key, int value) {
        bundle.putInt(key, value);
        return this;
    }

    public ActivityNavigator withData(String key, float value) {
        bundle.putFloat(key, value);
        return this;
    }

    public ActivityNavigator withData(String key, double value) {
        bundle.putDouble(key, value);
        return this;
    }

    public ActivityNavigator withData(String key, Serializable value) {
        bundle.putSerializable(key, value);
        return this;
    }

    public ActivityNavigator withData(String key, Parcelable value) {
        bundle.putParcelable(key, value);
        return this;
    }

    public ActivityNavigator withFillData(Functions.Func1<Void, Bundle> fillBundle) {
        fillBundle.execute(bundle);
        return this;
    }

    public ActivityNavigator configureIntent(Functions.Func1<Void, Intent> intentConfigurator) {
        this.intentConfigurator = intentConfigurator;
        return this;
    }

    private <T extends Activity> Intent getGoIntent(Class<T> activity) {
        Intent intent = new Intent(this.activity, activity);
        intent.putExtras(bundle);
        intent.addFlags(flags);
        if (intentConfigurator != null) intentConfigurator.execute(intent);
        return intent;
    }

    private void applyAnimation() {
        if (animationIn != -1 || animationOut != -1)
            this.activity.overridePendingTransition(animationIn, animationOut);
    }

    public <T extends Activity> void go(Class<T> activity) {
        go(activity, null);
    }

    public <T extends Activity> void go(Class<T> activity, Bundle options) {
        this.activity.startActivity(getGoIntent(activity), options);
        applyAnimation();
    }

    public <T extends Activity> void goForResult(Class<T> activity, int requestCode) {
        goForResult(activity, requestCode, null);
    }

    public <T extends Activity> void goForResult(Class<T> activity, int requestCode, Bundle options) {
        this.activity.startActivityForResult(getGoIntent(activity), requestCode, options);
        applyAnimation();
    }

    public void back() {
        activity.finish();
        applyAnimation();
    }

    public void backWithResult(int resultCode) {
        Intent intent = new Intent();
        intent.putExtras(bundle);
        activity.setResult(resultCode, intent);
        activity.finish();
    }

    public void finishActivities(int requestCode) {
        activity.finishActivity(requestCode);
    }
}

File: vkatz-lib/src/main/kotlin/by/vkatz/utils/ActivityNavigator.kt
package by.vkatz.utils

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable

import java.io.Serializable

/**
 * Created by Katz on 17.06.2016.
 */

@Suppress("MemberVisibilityCanPrivate")
class ActivityNavigator private constructor(private val activity: Activity) {
    companion object {

        fun forActivity(activity: Activity): ActivityNavigator = ActivityNavigator(activity)

        fun getData(activity: Activity): Bundle? {
            return try {
                activity.intent.extras
            } catch (e: Exception) {
                null
            }
        }
    }

    private var flags = 0
    private var animationIn = -1
    private var animationOut = -1
    private var finishCurrent = false
    private val bundle: Bundle = Bundle()
    private var options: Bundle? = null
    private var intentConfigurator: ((Intent) -> Unit)? = null

    fun finishCurrent(): ActivityNavigator {
        finishCurrent = true
        return this
    }

    fun noHistory(): ActivityNavigator {
        flags = flags or Intent.FLAG_ACTIVITY_NO_HISTORY
        return this
    }

    fun newTask(): ActivityNavigator {
        flags = flags or Intent.FLAG_ACTIVITY_NEW_TASK
        return this
    }

    fun clearTask(): ActivityNavigator {
        flags = flags or Intent.FLAG_ACTIVITY_CLEAR_TASK
        return this
    }

    fun clearTop(): ActivityNavigator {
        flags = flags or Intent.FLAG_ACTIVITY_CLEAR_TOP
        return this
    }

    fun renewTask(): ActivityNavigator {
        newTask()
        clearTask()
        clearTop()
        return this
    }

    fun reorderToFront(): ActivityNavigator {
        flags = flags or Intent.FLAG_ACTIVITY_REORDER_TO_FRONT
        return this
    }

    fun animation(animationIn: Int, animationOut: Int): ActivityNavigator {
        this.animationIn = animationIn
        this.animationOut = animationOut
        return this
    }

    fun options(options: Bundle): ActivityNavigator {
        this.options = options
        return this
    }

    fun withData(data: Bundle): ActivityNavigator {
        bundle.putAll(data)
        return this
    }

    fun withData(key: String, value: String): ActivityNavigator {
        bundle.putString(key, value)
        return this
    }

    fun withData(key: String, value: Int): ActivityNavigator {
        bundle.putInt(key, value)
        return this
    }

    fun withData(key: String, value: Float): ActivityNavigator {
        bundle.putFloat(key, value)
        return this
    }

    fun withData(key: String, value: Double): ActivityNavigator {
        bundle.putDouble(key, value)
        return this
    }

    fun withData(key: String, value: Serializable): ActivityNavigator {
        bundle.putSerializable(key, value)
        return this
    }

    fun withData(key: String, value: Parcelable): ActivityNavigator {
        bundle.putParcelable(key, value)
        return this
    }

    fun withFillData(fillBundle: (Bundle) -> Unit): ActivityNavigator {
        fillBundle(bundle)
        return this
    }

    fun configureIntent(intentConfigurator: (Intent) -> Unit): ActivityNavigator {
        this.intentConfigurator = intentConfigurator
        return this
    }

    private fun getGoIntent(activity: Class<out Activity>): Intent {
        val intent = Intent(this.activity, activity)
        intent.putExtras(bundle)
        intent.addFlags(flags)
        intentConfigurator?.invoke(intent)
        return intent
    }

    private fun applyAnimation() {
        if (animationIn != -1 || animationOut != -1)
            this.activity.overridePendingTransition(animationIn, animationOut)
    }

    fun <T : Activity> go(activity: Class<T>) {
        go(activity, null)
    }

    fun <T : Activity> go(activity: Class<T>, options: Bundle?) {
        this.activity.startActivity(getGoIntent(activity), options)
        applyAnimation()
        if (finishCurrent) this.activity.finish()
    }

    fun goForResult(activity: Class<out Activity>, requestCode: Int) {
        this.activity.startActivityForResult(getGoIntent(activity), requestCode, options)
        applyAnimation()
        if (finishCurrent) this.activity.finish()
    }

    fun back() {
        activity.finish()
        applyAnimation()
    }

    fun backWithResult(resultCode: Int) {
        val intent = Intent()
        intent.putExtras(bundle)
        activity.setResult(resultCode, intent)
        activity.finish()
    }

    fun finishActivities(requestCode: Int) {
        activity.finishActivity(requestCode)
    }
}
